File: Prog3/TrabalhoJAVA/src/app/LerCSV.java
package app;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class LerCSV {
    /**
     * Esse metodo le as entradas de um arquivo no formato CSV, cria e retorna uma
     * lista candidatos com base nas informacoes da entrada
     * 
     * @param diretorioArquivo string do caminho para o arquivo de entrada
     * @param date             data da eleicao no formato LocalDate
     * @return lista de candidatos
     */
    public static List<Candidato> lerCandidatos(String diretorioArquivo, LocalDate date) {
        List<Candidato> candidatos = new ArrayList<Candidato>();
        BufferedReader buffRead;
        int numCandidato, votosNominais, numPartido;
        String situacao, nomeCandidato, nomeUrna, destinoVoto;
        char sexo;
        LocalDate dataNasc = LocalDate.now();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");

        try {
            buffRead = new BufferedReader(new FileReader(diretorioArquivo));
            String linha = buffRead.readLine();

            while (true) {
                linha = buffRead.readLine();
                if (linha != null) {
                    String[] values = linha.split(",");
                    if (values[7].compareTo("Válido") != 0 || values.length < 9) {
                        continue;
                    }
                    numCandidato = Integer.parseInt(values[0]);
                    votosNominais = Integer.parseInt(values[1]);
                    situacao = values[2];
                    nomeCandidato = values[3];
                    nomeUrna = values[4];
                    sexo = values[5].charAt(0);
                    dataNasc = LocalDate.parse(values[6], formatter);
                    destinoVoto = values[7];
                    numPartido = Integer.parseInt(values[8]);
                    Candidato c = new Candidato(numCandidato, votosNominais, situacao, nomeCandidato, nomeUrna, sexo,
                            dataNasc, destinoVoto, numPartido, date);
                    candidatos.add(c);
                } else
                    break;
            }
            buffRead.close();

            return candidatos;

        } catch (FileNotFoundException e) {
            System.out.println("Arquivo de candidatos não encontrado");
            System.exit(0);
        } catch (IOException e) {
            System.out.println("Erro de IO no arquivo de candidato");
            System.exit(0);
        }

        return null;
    }

    /**
     * Esse metodo le as entradas de um arquivo no formato CSV, cria e retorna uma
     * lista de partidos com base nas informacoes da entrada
     * 
     * @param diretorioArquivo string do caminho para o arquivo de entrada
     * @param candidatos       lista com os candidatos da eleicao
     * @return lista de partidos
     */
    public static List<Partido> lerPartidos(String diretorioArquivo, List<Candidato> candidatos) {
        List<Partido> partidos = new ArrayList<Partido>();
        BufferedReader buffRead;
        int numPartido, votosLegenda;
        String nomePartido, siglaPartido;

        try {
            buffRead = new BufferedReader(new FileReader(diretorioArquivo));
            String linha = buffRead.readLine();

            while (true) {
                linha = buffRead.readLine();
                if (linha != null) {
                    String[] values = linha.split(",");
                    if (values.length < 4) {
                        continue;
                    }
                    numPartido = Integer.parseInt(values[0]);
                    votosLegenda = Integer.parseInt(values[1]);
                    nomePartido = values[2];
                    siglaPartido = values[3];
                    Partido p = new Partido(numPartido, votosLegenda, nomePartido, siglaPartido, candidatos);
                    partidos.add(p);
                } else
                    break;
            }
            buffRead.close();

            return partidos;

        } catch (FileNotFoundException e) {
            System.out.println("Arquivo de candidatos não encontrado");
            System.exit(0);
        } catch (IOException e) {
            System.out.println("Erro de IO no arquivo de candidato");
            System.exit(0);
        }

        return null;
    }
}

File: Prog3/TrabalhoC++/util/lerCSV.cpp
#include "lerCSV.h"

#include <stdlib.h>

#include <fstream>
#include <iostream>
#include <sstream>

#include "FileNotFoundException.h"

LerCSV::LerCSV() {
}

list<Candidato*> LerCSV::ler_candidatos(const string& diretorio_arquivo, const Date& date) {
    list<Candidato*> candidatos;
    int num_candidato, votos_nominais, num_partido;
    string situacao, nome_candidato, nome_urna, destino_voto;
    char sexo;
    string linha;
    Date data_nasc;
    string values[9];

    try {
        ifstream in(diretorio_arquivo);
        if (!in) throw FileNotFoundException(diretorio_arquivo);
        getline(in, linha);
        while (getline(in, linha)) {
            if (!linha.empty()) {
                stringstream check1(linha);
                string intermediate;
                for (int i = 0; getline(check1, intermediate, ','); i++)
                    values[i] = intermediate;

                num_candidato = stoi(values[0]);
                votos_nominais = stoi(values[1]);
                situacao = values[2];
                nome_candidato = values[3];
                nome_urna = values[4];
                sexo = values[5][0];
                data_nasc = Date(values[6]);
                destino_voto = values[7];
                if (destino_voto != "Válido")
                    continue;
                num_partido = stoi(values[8]);
                // cout << "Candidato " << count << ": " << num_candidato << " " << votos_nominais << " " << situacao << " " << nome_candidato << " " << nome_urna << " " << sexo << " " << data_nasc.to_string() << " " << destino_voto << " " << num_partido << endl;
                Candidato* c = new Candidato(num_candidato, votos_nominais, situacao, nome_candidato, nome_urna, sexo,
                                             data_nasc, destino_voto, num_partido, date);
                candidatos.push_back(c);
            } else
                break;
        }
        in.close();
    } catch (FileNotFoundException e) {
        cout << "Arquivo " << e.get_message() << " não encontrado" << endl;
        exit(1);
    }

    return candidatos;
}

list<Partido> LerCSV::ler_partidos(const string& diretorio_arquivo, const list<Candidato*>& candidatos) {
    list<Partido> partidos;
    int num_partido, votos_legenda;
    string nome_partido, sigla_partido;
    string linha;
    string values[4];

    try {
        ifstream in(diretorio_arquivo);
        if (!in) throw FileNotFoundException(diretorio_arquivo);
        getline(in, linha);
        while (getline(in, linha)) {
            if (!linha.empty()) {
                stringstream check1(linha);
                string intermediate;
                for (int i = 0; getline(check1, intermediate, ','); i++)
                    values[i] = intermediate;

                num_partido = stoi(values[0]);
                votos_legenda = stoi(values[1]);
                nome_partido = values[2];
                sigla_partido = values[3];
                // cout << "Partido " << count << ": " << num_partido << " " << votos_legenda << " " << nome_partido << " " << sigla_partido << endl;
                Partido p = Partido(num_partido, votos_legenda, nome_partido, sigla_partido, candidatos);
                partidos.push_back(p);
            } else
                break;
        }
        in.close();
    } catch (FileNotFoundException e) {
        cout << "Arquivo " << e.get_message() << " não encontrado" << endl;
        exit(1);
    }
    return partidos;
}

LerCSV::~LerCSV() {
}
